doc/source/user/building.rst

```diff
@@ -135,6 +138,18 @@ in the
 `meson_options.txt <https://github.com/numpy/numpy/blob/main/meson_options.txt>`_
 file.
 
+You can use OpenBLAS as a wheel, but must set up the build appropriately.
+There are two variants of the wheel, where possible we reccomend using the
+one with 64-bit integer interfaces, which allow for larger arrays::
+
+    pip install scipy-openblas64 spin
```

Comment:
Use of `spin` does not belong on this page, that should appear only on the pages aimed at contributors.

Reply:
Hmm. Above the user is directed to use `pip install -e . --no-build-isolation`. How woul d you suggest they get the needed modifications before running that:
- extract a properly configured `openblas.pc` from the scipy-openblas package
- point `PKG_CONFIG_PATH` to it
- modify `_distributor_init.py`

Reply:
I guess I could move the functionality I added to spin to do these things into to scipy-openblas, and then this would become
```
pip install scipy-openblas64
python -m scipy-openblas64 setup-repo="."
export PKG_CONFIG_PATH=${PWD}/.openblas   # bash
$env:PKG_CONFIG_PATH="$pwd/.openblas"  # powershell
pip install . -v -Csetup-args="--vsenv" -Csetup-args="-Duse-ilp64=true" -Csetup-args="-Dblas-symbol-suffix=64_"
```

Is that preferable?

Reply:
> How would you suggest they get the needed modifications before running that:

They don't, that really isn't all that relevant here. BLAS/LAPACK remain external and generic build time dependencies. The wheel is for us to depend on, not for them to mess with directly.

> Another idea: push all this into `numpy/meson.build`,

I'm not sure it's possible to set `PKG_CONFIG_PATH` from within Meson.

Either way, there may be an issue here in that the produced wheel will be cached by `pip` and will contain RPATH entries to a specific Python environment. That is going to break as soon as the user tries to install numpy into another environment.

I have to give it some more thought, but I think this is a little tricky, and perhaps a distraction from the main purpose of this PR right now. We want these wheels used by `numpy` and `scipy` wheels and in CI. I'd be inclined to scope this first PR to only that, and not user-facing changes.

Reply:
> the produced wheel will be cached by pip and will contain RPATH entries to a specific Python environment

I don't think that is a foregone conclusion. We can use the correct aduitwheel/delvehweel/delocate repair incantation, to `--exclude` the openblas runtimes and not do any RPATH mangling for `_multiarray_umath` and `_umath_linalg`. This works locally for me. That is why we need to modify `_disitributor_init` on all platforms to import the `libopenblas_python` shared object before importing numpy. That is why I am hung up on the runtime dependency on scipy-openblas. `pip install numpy` must install scipy-openblas, and I don't know how to express that dependency.

Reply:
> That is why I am hung up on the runtime dependency on scipy-openblas

Isn't that just adding it to `project.dependencies`?

Reply:
@andyfaff the problem is that "it" can be either nothing (if not using scipy-openblas), scipy-openblas32 or scipy-openblas64. I think @rgommers solution of mangling the `pyproject.toml` file in the before_build wheel script is the least bad option right now.

Reply:
There's still an issue here that I just thought of. A wheel technically is not allowed to contain dependencies that are not present in an sdist. This is because the PyPI API is per release rather than per artifact (https://warehouse.pypa.io/api-reference/json.html#release). This may be a hard blocker for now.

@mattip I think it's a good idea to split this PR up. There are 3 levels of changes here:
1. CI jobs (easiest and quite useful)
2. Depending on the new openblas wheels in our own releases
3. Allowing contributors and/or end users to use the openblas wheels

The bulk of the changes here are for (1), so it'd be great to get those merged. (2) is probably blocked for now; (3) can work but requires a bit more discussion on how to best do and document it.

Reply:
OK, I can concentrate on (1) for now.
> A wheel technically is not allowed to contain dependencies that are not present in an sdist

I don't understand. What dependencies are present in (which) wheel that are not present in an sdist?

Reply:
I'm guessing that on can't have different `[project.dependencies]` sections between the sdist and wheel?

Reply:
Indeed, and we cannot add the `scipy-openblasX` dependency to the sdist without breaking builds from source on platforms for which we don't have openblas wheels.

Reply:
We may be able to fix this with a metapackage or a do-nothing `scipy-openblasX` sdist, but it needs some further hack like that. Or a proper solution, like adding a "pypi provider" concept for external dependencies to https://peps.python.org/pep-0725/.

Reply:
Or could we combine both 32 bit and 64 into a single wheel, along with dummy do-nothing code?

Reply:
Ahh, now I get, sorry for being dense. If we add an ad-hoc dependency on, say, the manylinux2014 wheels for scipy-openblas64, and a dependency on, say, the 32-bit windows wheels for scipy-openblas32, and a dependency on, say, no openblas for macosx-arm64, then there is no way to reconcile all that for a single dependency for the sdist.

I guess another path would be to have a `numpy[openblas64]` wheel variant, but explaining `pip install numpy[openblas64]` seems like a nightmare.

> Or could we combine both 32 bit and 64 into a single wheel, along with dummy do-nothing code?

Hmm. The problem is of course the wheel size: users would be downloading at least one useless large shared object. That is why I made two wheels.

> PEP 725

I think the scipy-openblas wheels fall under the rejected category of [external dependencies which are also packaged on PyPI](https://peps.python.org/pep-0725/#specific-syntax-for-external-dependencies-which-are-also-packaged-on-pypi)

Reply:
> I think the scipy-openblas wheels fall under the rejected category of [external dependencies which are also packaged on PyPI](https://peps.python.org/pep-0725/#specific-syntax-for-external-dependencies-which-are-also-packaged-on-pypi)

Well I put it in that category because it seemed too niche - the PEP is still draft, so I can easily reverse that decision:)

> the manylinux2014 wheels for scipy-openblas64, and a dependency on, say, the 32-bit windows wheels for scipy-openblas32, and a dependency on, say, no openblas for macosx-arm64, then there is no way to reconcile all that for a single dependency for the sdist.

I think we don't need any 32-bit wheel for numpy at least (scipy is a different story), so it's a little easier. We could have platform selectors perhaps, determining whether we want `scipy-openblas64` or nothing.